RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1470 @@
+                            soiSamples.data(),
+                            mahi,
+                            recoParamsWithPS,
+                            configParameters.meanTime,
+                            configParameters.timeSigmaSiPM,
+                            configParameters.timeSigmaHPD,
+                            configParameters.applyTimeSlew,
+                            configParameters.tzeroTimeSlew,
+                            configParameters.slopeTimeSlew,
+                            configParameters.tmaxTimeSlew);
+
+        uint32_t threadsPerBlock = configParameters.kernelMinimizeThreads[0];
+        auto blocks_1d = cms::alpakatools::divide_up_by(totalChannels, threadsPerBlock);
+
+        auto workDivPrep1D = cms::alpakatools::make_workdiv<Acc1D>(blocks_1d, threadsPerBlock);
+
+        alpaka::exec<Acc1D>(queue,
+                            workDivPrep1D,
+                            mahi::Kernel_minimize<8, 8>{},
+                            outputGPU,
+                            amplitudes.data(),
+                            pulseMatrices.data(),
+                            pulseMatricesM.data(),
+                            pulseMatricesP.data(),
+                            mahiPulseOffsets,
+                            noiseTerms.data(),
+                            electronicNoiseTerms.data(),
+                            soiSamples.data(),
+                            mahi,
+                            configParameters.useEffectivePedestals,
+                            f01HEDigis,
+                            f5HBDigis,
+                            f3HBDigis);
+      }
+
+    }  // namespace reconstruction
+  }    // namespace hcal
+}  // namespace ALPAKA_ACCELERATOR_NAMESPACE
+
+namespace alpaka::trait {
+  using namespace ALPAKA_ACCELERATOR_NAMESPACE::hcal::reconstruction::mahi;
+
+  //! The trait for getting the size of the block shared dynamic memory for Kernel_prep_1d_and_initialize.
+  template <typename TAcc>
+  struct BlockSharedMemDynSizeBytes<Kernel_prep1d_sameNumberOfSamples, TAcc> {
+    //! \return The size of the shared memory allocated for a block.
+    template <typename TVec, typename... TArgs>
+    ALPAKA_FN_HOST_ACC static auto getBlockSharedMemDynSizeBytes(Kernel_prep1d_sameNumberOfSamples const&,
+                                                                 TVec const& threadsPerBlock,
+                                                                 TVec const& elemsPerThread,
+                                                                 TArgs const&...) -> std::size_t {
+      // return the amount of dynamic shared memory needed
+
+      // threadsPerBlock[1] = threads2d.x = windowSize = 8
+      // threadsPerBlock[0] = threads2d.y = nchannels_per_block = 32
+      // elemsPerThread = 1
+      std::size_t bytes =
+          threadsPerBlock[0u] * elemsPerThread[0u] * ((2 * threadsPerBlock[1u] + 2) * sizeof(float) + sizeof(uint64_t));
```

Comment:
Is `elemsPerThread[1u]` not needed ?

Reply:
I think this can be
```
std::size_t bytes =
          threadsPerBlock[0u] * elemsPerThread[0u] * ((2 * threadsPerBlock[1u] * elemsPerThread[1u]+ 2) * sizeof(float) + sizeof(uint64_t));
```

Reply:
ok

Reply:
Done in 4bcedbfa3762f46753bcb6564102898708f5562b